index.d.ts

```diff
@@ -0,0 +1,116 @@
+import { DocumentClient } from "aws-sdk/clients/dynamodb";
+
+type FieldTypes =
+  | "string"
+  | "boolean"
+  | "number"
+  | "list"
+  | "map"
+  | "binary"
+  | "set";
+
+type CoerceTypes = "string" | "boolean" | "number" | "list";
+
+type SetTypes = "string" | "number" | "binary";
+
+type CompositeKeyFieldConfiguration<T> = FieldDefinition<T> & {
+  save?: boolean;
+};
+
+type CompositeKeyFieldDefinition<T> = [
+  string,
+  number,
+  CompositeKeyFieldConfiguration<T>?
+];
+
+interface FieldDefinition<T> {
+  type?: FieldTypes;
+  coerce?: CoerceTypes;
+  default?: (data: T) => any;
+  onUpdate?: boolean;
+  hidden?: boolean;
+  required?: boolean | "always";
+  alias?: keyof T;
+  setType?: SetTypes;
+}
+
+type SchemaDefinition<T> = Record<
+  string,
+  FieldTypes | FieldDefinition<T> | CompositeKeyFieldDefinition<T>
+>;
+
+interface ModelDefinition<T> {
+  table: string;
+  partitionKey: string;
+  sortKey?: string;
+  model?: boolean;
+  timestamps?: boolean;
+  created?: string;
+  modified?: string;
+  schema: SchemaDefinition<T>;
+}
+
+type UpdateNumberOperations =
+  | {
+      $add: number;
+    }
+  | number;
+
+type UpdateArrayOperations<T> =
+  | {
+      $add?: T[];
+      $delete?: T[];
+      $append?: T[];
+      $prepend?: T[];
+      $remove?: T[];
+      [index: number]: T;
+    }
+  | T[];
+
+type UpdateItemObjectField<T> =
+  | {
+      $set:
+        | {
+            [P in keyof T]?: UpdateItemField<T[P]>;
+          }
+        | {
+            [index: string]: any;
+          };
+    }
+  | T;
+
+type UpdateItemField<T> = T extends Array<infer P>
+  ? UpdateArrayOperations<P>
+  : T extends object
+  ? UpdateItemObjectField<T>
+  : T extends number
+  ? UpdateNumberOperations
+  : T;
+
+type UpdateItemWrapper<T> = { [P in keyof T]?: UpdateItemField<T[P]> };
+
+export class Model<T> {
```

Comment:
Looks like all your return values on the Dynamo Operations return the input instead of the output. Assuming the `DocumentClient` has the return values of these methods typed as `Output` instead. Although I could be wrong.

e.g. `DocumentClient.PutItemInput` -> `DocumentClient.PutItemOutput`

Reply:
Hey! This is due to the way this library works. It actually does not do the db operation for you. It just creates the input parameters for the `DocumentClient` method which you need to call manually.  Hence why all the outputs are `ItemInput`s.

Reply:
Also, lots are changing in v0.2 to accommodate the feedback from the original version, so this will need to change anyway.

Reply:
> Hey! This is due to the way this library works. It actually does not do the db operation for you. It just creates the input parameters for the `DocumentClient` method which you need to call manually. Hence why all the outputs are `ItemInput`s.

Ahh makes sense. Reading through things to quickly. My fault.